wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CacheConnection.java

```diff
@@ -257,10 +328,197 @@ private void returnConnectionBackToPool(StatefulRedisConnection <byte[], byte[]>
+        !CACHE_IAM_USERNAME.getString(this.connectionProperties).trim().isEmpty();
+  }
+
+  private boolean hasTraditionalAuthentication() {
+    return CACHE_USERNAME.getString(this.connectionProperties) != null &&
+        CACHE_PASSWORD.getString(this.connectionProperties) != null;
+  }
+
+  private void refreshIamTokenIfExpired() {
+    String username = CACHE_IAM_USERNAME.getString(this.connectionProperties);
+    String region = CACHE_IAM_REGION.getString(this.connectionProperties);
+    String hostname = this.hostAndPort[0];
+    int port = this.hostAndPort.length > 1 ? Integer.parseInt(this.hostAndPort[1]) : 6379;
+
+    if (username == null || hostname == null || region == null) {
+      throw new RuntimeException("Missing required IAM authentication properties");
+    }
+
+    String cacheKey = IamAuthUtils.getCacheKey(username, hostname, port, Region.of(region));
+    TokenInfo tokenInfo = IamAuthCacheHolder.getCachedToken(cacheKey);
+    if (tokenInfo == null || tokenInfo.isExpired()) {
+      ElastiCacheIamTokenUtility tokenUtility = new ElastiCacheIamTokenUtility();
+
+      String cacheName = CACHE_NAME.getString(this.connectionProperties);
+      if (cacheName != null) {
+        tokenUtility.setCacheName(cacheName);
+        LOGGER.finest("Using ElastiCache cache name: " + cacheName);
+      }
+      String token = tokenUtility.generateAuthenticationToken(
+          AwsCredentialsManager.getProvider(null, this.connectionProperties),
+          Region.of(region),
+          hostname,
+          port,
+          username);
+
+      connectionProperties.setProperty("iamToken", token);
+      LOGGER.finest("[" + new java.util.Date() + "] Generated fresh IAM token for ElastiCache");
+
+      // Cache the new token
+      Instant expiration = Instant.now().plus(14, ChronoUnit.MINUTES);
+      IamAuthCacheHolder.cachePutToken(cacheKey, new TokenInfo(token, expiration));
+    } else {
+      connectionProperties.setProperty("iamToken", tokenInfo.getToken());
+    }
+  }
+
   // Used for unit testing only
   protected void setConnectionPools(GenericObjectPool<StatefulRedisConnection<byte[], byte[]>> readPool,
       GenericObjectPool<StatefulRedisConnection<byte[], byte[]>> writePool) {
     readConnectionPool = readPool;
     writeConnectionPool = writePool;
   }
+
+  private void setUpAuthenticationProperties(Properties properties) {
+    boolean iamAuthEnabled = CACHE_IAM_USERNAME.getString(properties) != null &&
+        !CACHE_IAM_USERNAME.getString(properties).trim().isEmpty();
+    if (iamAuthEnabled) {
+      setUpIamAuthentication(properties);
+    } else if (hasTraditionalCredentials(properties)) {
+      setupTraditionalAuthentication(properties);
+    }
+  }
+
+  private void setUpIamAuthentication(Properties properties) {
+    String hostname = extractHostnameFromCacheEndpoint(properties);
+    if (hostname != null && isElastiCacheEndpoint(hostname)) {
```

Comment:
Remove the check for `isElastiCacheEndpoint()`. We should NOT have elasticache specific logic here.